subsys/bluetooth/services/ots/ots_dir_list.c

```diff
@@ -20,13 +20,30 @@
 
 LOG_MODULE_DECLARE(bt_ots, CONFIG_BT_OTS_LOG_LEVEL);
 
+/* In order to transmit the maximum size we choose CONFIG_BT_OTS_L2CAP_CHAN_TX_MTU
+ * as the buffer size. However, directory listing encoding can only happen at
+ * directory entry boundary so we must ensure the MTU value is rounded up to the
+ * nearest directory entry size to not overflow our buffer. Also since we can only
+ * encode at object entry boundary, we need to make sure we have at least room for
+ * two entries so entries can be read from the middle.
+ */
+#define OTS_DIR_LIST_BUFFER_SIZE MAX(2 * DIR_LIST_OBJ_RECORD_MAX_SIZE, \
+	DIR_LIST_OBJ_RECORD_MAX_SIZE * \
+	ceiling_fraction(CONFIG_BT_OTS_L2CAP_CHAN_TX_MTU, DIR_LIST_OBJ_RECORD_MAX_SIZE))
+
+struct bt_ots_dir_list {
+	struct net_buf_simple net_buf;
+	struct bt_gatt_ots_object *dir_list_obj;
+	uint8_t _content[OTS_DIR_LIST_BUFFER_SIZE];
+};
+
 static struct bt_ots_dir_list dir_lists[CONFIG_BT_OTS_MAX_INST_CNT];
 
-static void dir_list_object_encode(struct bt_gatt_ots_object *obj,
-				   struct net_buf_simple *net_buf)
+static size_t dir_list_object_encode(const struct bt_gatt_ots_object *obj,
+				     struct net_buf_simple *net_buf)
 {
 	uint8_t flags = 0;
-	uint8_t *start;
+	uint8_t *start = NULL;
```

Comment:
Can you document the behavior on when `net_buf` may be NULL? 

To me it looks like that if `dir_list_object_encode` is called with `net_buf == NULL`, then it will just get the size of the directory listing. Is that correct? If so, can we make that more clear for future readers?

Reply:
Given the simplicity of the function, perhaps we should just consider making a new function to get the size instead like `dir_list_get_size`

Reply:
> Can you document the behavior on when net_buf may be NULL?
>
> To me it looks like that if dir_list_object_encode is called with net_buf == NULL, then it will just get the size of the directory listing. Is that correct? If so, can we make that more clear for future readers?

This is correct, I will add documentation to this effect.

> Given the simplicity of the function, perhaps we should just consider making a new function to get the size instead like dir_list_get_size

I am not sure I understand what you mean. If the suggestion is to create a separate function that doesn't have all the `if (net_buf != NULL)`, I am not sure I like that idea since it creates two places where an Object Record format is known.

Reply:
>  I am not sure I like that idea since it creates two places where an Object Record format is known.

I am not sure what you mean by this. Can you elaborate? 

My thought was that if we call this function with `net_buf == NULL` just to get the size, I think it would be better suited as a new function.

Reply:
The object record size is variable based on both name length of the object and presence of some optional fields. If we have a function that just returns the size of the record and a function that encodes the record, then we would have the logic of what content builds up a record in two places. The approach taken here was to keep the logic of what makes up a record in a single place.

Reply:
> then we would have the logic of what content builds up a record in two places

Agreed, but I don't see that as a problem though. 
I'd rather have 2 functions that can work on OTS object metadata, rather than having a stateless version (getting length) and a statefull version (encoding data) depending on whether an argument is NULL or not.